File: programmers/kakao/수식최대화.py
# 프로그래머스 수식 최대화
# 21.05.07

import sys
from collections import deque
from itertools import permutations
import copy
rd = sys.stdin.readline

def calculate(arr, priority):
    q1 = deque()
    q2 = deque()
    q1 = copy.deepcopy(arr)
    i = 0
    while i < 3:
        while q1:
            if q2 and q2[-1] == priority[i]:
                q2.pop()
                v1 = int(q2.pop())
                v2 = int(q1.popleft())
                if priority[i] == "*":
                    q2.append(str(v1 * v2))
                elif priority[i] == "+":
                    q2.append(str(v1 + v2))
                else:
                    q2.append(str(v1 - v2))
            else:
                q2.append(q1.popleft())
        q1 = copy.deepcopy(q2)
        q2.clear()
        i += 1
    
    return int(q1[0])



def solution(expression):
    answer = 0
    exp = deque()
    operand = ""
    for s in expression:
        if s == "*" or s == "+" or s == "-":
            exp.append(operand)
            operand = ""
            exp.append(s)
        else:
            operand += s
    exp.append(operand)

    priorities = list(permutations(["*", "+", "-"], 3))

    for p in priorities:
        temp = abs(calculate(exp, p))
        if temp > answer:
            answer = temp

    return answer


print(solution("100-200*300-500+20"))
print(solution("50*6-3*2"))

File: programmers/kakao/수식최대화.swift
// 프로그래머스 수식 최대화
// 21.05.07

import Foundation

func permutation<T>(_ arr: [T], _ r: Int) -> [[T]] {
  var result = [[T]]()
  var preElements = [T]()

  func dfs(_ elements: [T]) {
    if elements.count == arr.count - r {
      result.append(preElements)
    }

    for i in 0..<elements.count {
      var nextElements = elements
      nextElements.remove(at: i)

      preElements.append(elements[i])
      dfs(nextElements)
      preElements.removeLast()
    }
  }
  dfs(arr)
  return result
}

func solution(_ expression:String) -> Int64 {
  let operators = ["*", "+", "-"]
  var priorities = permutation(operators, 3)
  var expOperands = [Int64]()
  var expOperators = [String]()
  var str = ""
  var answer: Int64 = 0

  // 1: 연산자와 피연산자 분리
  for char in expression {
    if operators.contains(String(char)) {
      expOperators.append(String(char))
      expOperands.append(Int64(str)!)
      str = ""
    } else {
      str += String(char)
    }
  }
  expOperands.append(Int64(str)!)

  // 2: 모든 우선순위에 대해 수행
  for priority in priorities {
    var tempOperands = expOperands
    var tempOperators = expOperators

    for op in priority {
      var i = 0
      while i < tempOperators.count {
        if tempOperators[i] == op {
          var value: Int64
          if op == "*" {
            value = tempOperands[i] * tempOperands[i + 1]
          } else if op == "+" {
            value = tempOperands[i] + tempOperands[i + 1]
          } else {
            value = tempOperands[i] - tempOperands[i + 1]
          }
          tempOperands[i] = value
          tempOperands.remove(at: i + 1)
          tempOperators.remove(at: i)
          i -= 1
        }
        i += 1
      }
    }
    if abs(tempOperands[0]) > answer {
      answer = abs(tempOperands[0])
    }
  }

  return answer
}

print(solution("100-200*300-500+20"))
print(solution("50*6-3*2"))